### PR TITLE
Distributed AI merge

### PR DESCRIPTION
**Pull request by mateicristea88**
Adding distributed AI code to superapp. Adds an app in the IPv8 menu. Work is still being done locally, without communication between peers, but the learning algorithm can only access current datapoints (as a real world product would do)

### CODE REVIEW

**Comment by MattSkala** [comment]
Thank for the PR.

I just quickly tested the app, and I think it would help to explain what is the valid range of values for different fields and perform some basic validation. E.g., the app crashes if you enter x = 1, y = 2, x = 3, y = 4, iterations = 5. Attached is the stack trace.

```
    Process: nl.tudelft.trustchain, PID: 31571
    java.lang.IllegalStateException: Could not execute method for android:onClick
        at androidx.appcompat.app.AppCompatViewInflater$DeclaredOnClickListener.onClick(AppCompatViewInflater.java:402)
        at android.view.View.performClick(View.java:7259)
        at android.view.View.performClickInternal(View.java:7236)
        at android.view.View.access$3600(View.java:801)
        at android.view.View$PerformClick.run(View.java:27892)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Method.invoke(Native Method)
        at androidx.appcompat.app.AppCompatViewInflater$DeclaredOnClickListener.onClick(AppCompatViewInflater.java:397)
        at android.view.View.performClick(View.java:7259) 
        at android.view.View.performClickInternal(View.java:7236) 
        at android.view.View.access$3600(View.java:801) 
        at android.view.View$PerformClick.run(View.java:27892) 
        at android.os.Handler.handleCallback(Handler.java:883) 
        at android.os.Handler.dispatchMessage(Handler.java:100) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7356) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930) 
     Caused by: java.lang.ArithmeticException: divide by zero
        at nl.tudelft.trustchain.distributedAI.java.Entity.perform(Entity.java:123)
        at nl.tudelft.trustchain.distributedAI.java.Entity.run(Entity.java:67)
        at nl.tudelft.trustchain.distributedAI.DistributedActivity.sendMessage(DistributedActivity.java:53)
```

You could also add a simple README.md to your module to explain how it works and add a link to the root /README.md (see other modules for inspiration).

**Comment by mateicristea88** [comment]
> Thank for the PR.
> 
> I just quickly tested the app, and I think it would help to explain what is the valid range of values for different fields and perform some basic validation. E.g., the app crashes if you enter x = 1, y = 2, x = 3, y = 4, iterations = 5. Attached is the stack trace.
> 
> ```
>     Process: nl.tudelft.trustchain, PID: 31571
>     java.lang.IllegalStateException: Could not execute method for android:onClick
>         at androidx.appcompat.app.AppCompatViewInflater$DeclaredOnClickListener.onClick(AppCompatViewInflater.java:402)
>         at android.view.View.performClick(View.java:7259)
>         at android.view.View.performClickInternal(View.java:7236)
>         at android.view.View.access$3600(View.java:801)
>         at android.view.View$PerformClick.run(View.java:27892)
>         at android.os.Handler.handleCallback(Handler.java:883)
>         at android.os.Handler.dispatchMessage(Handler.java:100)
>         at android.os.Looper.loop(Looper.java:214)
>         at android.app.ActivityThread.main(ActivityThread.java:7356)
>         at java.lang.reflect.Method.invoke(Native Method)
>         at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
>         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
>      Caused by: java.lang.reflect.InvocationTargetException
>         at java.lang.reflect.Method.invoke(Native Method)
>         at androidx.appcompat.app.AppCompatViewInflater$DeclaredOnClickListener.onClick(AppCompatViewInflater.java:397)
>         at android.view.View.performClick(View.java:7259) 
>         at android.view.View.performClickInternal(View.java:7236) 
>         at android.view.View.access$3600(View.java:801) 
>         at android.view.View$PerformClick.run(View.java:27892) 
>         at android.os.Handler.handleCallback(Handler.java:883) 
>         at android.os.Handler.dispatchMessage(Handler.java:100) 
>         at android.os.Looper.loop(Looper.java:214) 
>         at android.app.ActivityThread.main(ActivityThread.java:7356) 
>         at java.lang.reflect.Method.invoke(Native Method) 
>         at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492) 
>         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930) 
>      Caused by: java.lang.ArithmeticException: divide by zero
>         at nl.tudelft.trustchain.distributedAI.java.Entity.perform(Entity.java:123)
>         at nl.tudelft.trustchain.distributedAI.java.Entity.run(Entity.java:67)
>         at nl.tudelft.trustchain.distributedAI.DistributedActivity.sendMessage(DistributedActivity.java:53)
> ```
> 
> You could also add a simple README.md to your module to explain how it works and add a link to the root /README.md (see other modules for inspiration).

Good catch. I still need to some some input validation, there are some division by zero errors in case our error is too low. I'll take care of it.

I made a readme specific to our app, I'll be sure to also reference it in the main readme

**Comment by mateicristea88** [comment]
Fixed the data validation issues, added readme. Couldn't figure out how to squash the commits together, sorry if it looks ugly.

File: settings.gradle [hunk 1]
```diff
@@ -13,6 +13,8 @@ include ':debug'
 include ':trustchain-trader'
 include ':trustchain-payloadgenerator'
 include ':currencyii'
+include ':FOC'
```
Comment (MattSkala) [inline settings.gradle:16]: Probably an extra line here, FOC module has been renamed to freedomOfComputing

File: distributedai/docs/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# Distributed-AI
+
+Luxury communism is an Android application built on top of [IPv8](https://github.com/Tribler/kotlin-ipv8)
```
Comment (MattSkala) [inline distributedai/docs/README.md:3]: Replace "luxury communism" with "distributed AI"? :)

Comment (mateicristea88) [reply distributedai/docs/README.md:3]: oops

File: settings.gradle (hunk 1 above)
Comment (mateicristea88) [reply settings.gradle:16]: fixed, thanks

**Comment by mateicristea88** [comment]
Should be resolved now, thanks for looking over it.

**Pull request merged**